Flaky SSH cloning failure in tests

Extracted from https://github.com/datalad/datalad/pull/6550#issuecomment-1073024568:

This only happens on AppVeyor (macOS again):

```
======================================================================
ERROR: datalad.core.distributed.tests.test_clone.test_ria_postclonecfg('ssh://datalad-test:/Users/appveyor/DLTMP/datalad_temp_ix8umpb9', '07c27167-6fef-443c-bbb7-3eec35daddc3')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/appveyor/venv3.8.12/lib/python3.8/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/Users/appveyor/projects/datalad/datalad/tests/utils.py", line 288, in _wrap_skip_ssh
    return func(*args, **kwargs)
  File "/Users/appveyor/projects/datalad/datalad/tests/utils.py", line 874, in _wrap_with_tempfile
    return t(*(arg + (filename,)), **kw)
  File "/Users/appveyor/projects/datalad/datalad/tests/utils.py", line 874, in _wrap_with_tempfile
    return t(*(arg + (filename,)), **kw)
  File "/Users/appveyor/projects/datalad/datalad/core/distributed/tests/test_clone.py", line 958, in _test_ria_postclonecfg
    riaclone = clone('ria+{}#{}'.format(url, dsid), clone_path)
  File "/Users/appveyor/projects/datalad/datalad/interface/utils.py", line 447, in eval_func
    return return_func(*args, **kwargs)
  File "/Users/appveyor/projects/datalad/datalad/interface/utils.py", line 439, in return_func
    results = list(results)
  File "/Users/appveyor/projects/datalad/datalad/interface/utils.py", line 424, in generator_func
    raise IncompleteResultsError(
datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully. 1 failed:
[{'action': 'install',
  'message': ('Failed to clone from any candidate source URL. Encountered '
              'errors per each url were:\n'
              '- %s',
              'ssh://datalad-test/Users/appveyor/DLTMP/datalad_temp_ix8umpb9/07c/27167-6fef-443c-bbb7-3eec35daddc3\n'
              "  CommandError: 'git -c diff.ignoreSubmodules=none clone "
              '--progress '
              'ssh://datalad-test/Users/appveyor/DLTMP/datalad_temp_ix8umpb9/07c/27167-6fef-443c-bbb7-3eec35daddc3 '
              "/Users/appveyor/DLTMP/datalad_temp__test_ria_postclonecfgw5zk_49p' "
              "failed with exitcode 128 [err: 'Cloning into "
              "'/Users/appveyor/DLTMP/datalad_temp__test_ria_postclonecfgw5zk_49p'...\n"
              '\r'
              'remote: Total 37 (delta 7), reused 0 (delta 0)        \n'
              "fatal: remote transport reported error']"),
  'path': '/Users/appveyor/DLTMP/datalad_temp__test_ria_postclonecfgw5zk_49p',
  'source_url': 'ria+ssh://datalad-test:/Users/appveyor/DLTMP/datalad_temp_ix8umpb9#07c27167-6fef-443c-bbb7-3eec35daddc3',
  'status': 'error',
  'type': 'dataset'}]
```

This seems flaky. Logging into that AppVeyor build, showed, that this happens at different spots in this test.
Sometimes this clone seems to work out fine but then the subsequent `get` on a subdataset fails the same way.
So, currently the failure happens at line 958 in `test_clone.py` and on previous run (exact same commit) it only failed at line 1017.
Moreover, this should not be the only test where we clone from RIA via SSH. Not clear to me yet, how this is one is different.

Looking into this, I am seeing a Broken Pipe Error:
```
[DEBUG] ...>runner:192  Finished ['ssh', '-o', 'ControlPath=/Users/appveyor/Library/Caches/datalad/sockets/fb3f4327', '-o', 'SendEnv=GIT_PROTOCOL', 'datalad-test', "git-upload-pack '/private/var/folders/5s/g225f6nd6jl4g8tshbh1ltk40000gn/T/datalad_temp_vlst5whp/376/1c829-d43c-420a-95fb-4467944477c4'"] with status 0 
[ERROR] ...>main:136,185  [Errno 32] Broken pipe (BrokenPipeError) 
fatal: remote transport reported error']
```

And `git-upload-pack` seems a bit off indeed:

```
appveyor$ git-upload-pack '/private/var/folders/5s/g225f6nd6jl4g8tshbh1ltk40000gn/T/datalad_temp_vlst5whp/376/1c829-d43c-420a-95fb-4467944477c4'
010d74b31e4b1f6a81373783c6520507909436ca0f3b HEADmulti_ack thin-pack side-band side-band-64k ofs-delta shallow deepen-since deepen-not deepen-relative no-progress include-tag multi_ack_detailed symref=HEAD:refs/heads/dl-test-branch object-format=sha1 agent=git/2.35.1
004774b31e4b1f6a81373783c6520507909436ca0f3b refs/heads/dl-test-branch
0000
```
*hanging at this point*

And, of course, there's no problem running this right afterwards:
```
appveyor$ datalad clone "ria+ssh://datalad-test/private/var/folders/5s/g225f6nd6jl4g8tshbh1ltk40000gn/T/datalad_temp_vlst5whp#3761c829-d43c-420a-95fb-4467944477c4" test5                                                                 
Clone attempt:   0%|                                                                                                                                                                                                              | 0.00/1.00 [00:00<?, ? Candidate locations/s]@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@    WARNING: REMOTE HOST IDENTIFICATION HAS CHANGED!     @
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
IT IS POSSIBLE THAT SOMEONE IS DOING SOMETHING NASTY!
Someone could be eavesdropping on you right now (man-in-the-middle attack)!
It is also possible that a host key has just been changed.
The fingerprint for the ECDSA key sent by the remote host is
SHA256:b7Q9hN2pEJGEvu/BlO2GUD/EV+H/xlmDqx7oCUosGbg.
Please contact your system administrator.
Add correct host key in /Users/appveyor/.ssh/known_hosts to get rid of this message.
Offending ECDSA key in /Users/appveyor/.ssh/known_hosts:154
Password authentication is disabled to avoid man-in-the-middle attacks.
Keyboard-interactive authentication is disabled to avoid man-in-the-middle attacks.
install(ok): /Users/appveyor/projects/test5 (dataset) 
```

--- a/changelog.d/pr-7072.md
+++ b/changelog.d/pr-7072.md
@@ -1,0 +1,6 @@
+### Bug Fixes
+
+- Avoid writing to stdout/stderr from within datalad sshrun. This could lead to
+  broken pipe errors when cloning via SSH and was superfluous to begin with.
+  Fixes https://github.com/datalad/datalad/issues/6599 via
+  https://github.com/datalad/datalad/pull/7072 (by @bpoldrack)

--- a/datalad/core/distributed/tests/test_clone.py
+++ b/datalad/core/distributed/tests/test_clone.py
@@ -1158,7 +1158,6 @@ def _postclonetest_prepare(lcl, storepath, storepath2, link):
 
 
 # TODO?: make parametric again on _test_ria_postclonecfg
-@known_failure_osx  # https://github.com/datalad/datalad/issues/6599
 @known_failure_windows  # https://github.com/datalad/datalad/issues/5134
 @slow  # 14 sec on travis
 def test_ria_postclonecfg():

--- a/datalad/support/sshconnector.py
+++ b/datalad/support/sshconnector.py
@@ -174,11 +174,13 @@ class BaseSSHConnection(object):
           (probably most) of the available configuration options should not be
           set here because they can critically change the properties of the
           connection. This exists to allow options like SendEnv to be set.
+        log_output: bool
+          Whether to capture and return stdout+stderr.
 
         Returns
         -------
         tuple of str
-          stdout, stderr of the command run.
+          stdout, stderr of the command run, if `log_output` was `True`
         """
         raise NotImplementedError
 

--- a/datalad/support/sshrun.py
+++ b/datalad/support/sshrun.py
@@ -116,10 +116,13 @@ class SSHRun(Interface):
         # use an empty temp file as stdin if none shall be connected
         stdin_ = tempfile.TemporaryFile() if no_stdin else sys.stdin
         try:
-            out, err = ssh(cmd, stdin=stdin_, log_output=False,
-                           options=options)
+            # We pipe the SSH process' stdout/stderr by means of
+            # `log_output=False`. That's necessary to let callers - for example
+            # git-clone - communicate with the SSH process. Hence, we expect no
+            # output being returned from this call:
+            out, err = ssh(cmd, stdin=stdin_, log_output=False, options=options)
+            assert not out
+            assert not err
         finally:
             if no_stdin:
                 stdin_.close()
-        os.write(1, out.encode('UTF-8'))
-        os.write(2, err.encode('UTF-8'))

